File: examples/go/order-specification/order/specification.go
package order

import "time"

type Specification interface {
	IsSatisfiedBy(candidate Order) bool
}

type PaidSpecification struct{}

func (PaidSpecification) IsSatisfiedBy(candidate Order) bool {
	return candidate.IsPaid()
}

type NotRefundedSpecification struct{}

func (NotRefundedSpecification) IsSatisfiedBy(candidate Order) bool {
	return !candidate.IsRefunded()
}

type RefundWindowOpenSpecification struct {
	Now time.Time
}

func (s RefundWindowOpenSpecification) IsSatisfiedBy(candidate Order) bool {
	return s.Now.Before(candidate.RefundCutoff())
}

type andSpecification struct {
	specifications []Specification
}

func And(specifications ...Specification) Specification {
	return andSpecification{specifications: append([]Specification(nil), specifications...)}
}

func (s andSpecification) IsSatisfiedBy(candidate Order) bool {
	for _, specification := range s.specifications {
		if !specification.IsSatisfiedBy(candidate) {
			return false
		}
	}

	return true
}

File: examples/typescript/order-specification/order/specification.ts
import { Order } from "./order";

export interface Specification<T> {
  isSatisfiedBy(candidate: T): boolean;
}

export class PaidSpecification implements Specification<Order> {
  isSatisfiedBy(candidate: Order): boolean {
    return candidate.isPaid();
  }
}

export class NotRefundedSpecification implements Specification<Order> {
  isSatisfiedBy(candidate: Order): boolean {
    return !candidate.isRefunded();
  }
}

export class RefundWindowOpenSpecification implements Specification<Order> {
  constructor(private readonly now: Date) {}

  isSatisfiedBy(candidate: Order): boolean {
    return this.now < candidate.refundCutoff();
  }
}

class AndSpecification<T> implements Specification<T> {
  constructor(private readonly specifications: Specification<T>[]) {}

  isSatisfiedBy(candidate: T): boolean {
    for (const specification of this.specifications) {
      if (!specification.isSatisfiedBy(candidate)) {
        return false;
      }
    }

    return true;
  }
}

export function andSpecification<T>(...specifications: Specification<T>[]): Specification<T> {
  return new AndSpecification(specifications);
}
